12-offer-encoding.md

```diff
@@ -0,0 +1,870 @@
+`messagename` is the name of the TLV stream being signed (i.e. "invoice_request" or "invoice") and the `fieldname` is the TLV field containing the
+signature (e.g. "signature").
+
+The formulation of the Merkle tree is similar to that proposed in
+[BIP-341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki),
+with each TLV leaf paired with a nonce leaf to avoid
+revealing adjacent nodes in proofs.
+
+The Merkle tree's leaves are, in TLV-ascending order for each tlv:
+1. The H("LnLeaf",tlv).
+2. The H("LnNonce"||first-tlv,tlv-type) where first-tlv is the numerically-first TLV entry in the stream, and tlv-type is the "type" field (1-9 bytes) of the current tlv.
+
+The Merkle tree inner nodes are H("LnBranch", lesser-SHA256||greater-SHA256);
+this ordering means proofs are more compact since left/right is
+inherently determined.
+
+If there is not exactly a power of 2 leaves, then the tree depth will
+be uneven, with the deepest tree on the lowest-order leaves.
+
+e.g. consider the encoding of an `invoice` `signature` with TLVs TLV1, TLV2, and TLV3 (of types 1, 2 and 3 respectively):
+
+```
+L1=H("LnLeaf",TLV1)
+L1nonce=H("LnNonce"||TLV1,1)
+L2=H("LnLeaf",TLV2)
+L2nonce=H("LnNonce"||TLV1,2)
+L3=H("LnLeaf",TLV3)
+L3nonce=H("LnNonce"||TLV1,3)
+
+Assume L1 < L1nonce, L2 > L2nonce and L3 > L3nonce.
+
+   L1    L1nonce                      L2   L2nonce                L3   L3nonce
+     \   /                             \   /                       \   /
+      v v                               v v                         v v
+L1A=H("LnBranch",L1||L1nonce) L2A=H("LnBranch",L2nonce||L2)  L3A=H("LnBranch",L3nonce||L3)
+                 
+Assume L1A < L2A:
+
+       L1A   L2A                                 L3A=H("LnBranch",L3nonce||L3)
+         \   /                                    |
+          v v                                     v
+  L1A2A=H("LnBranch",L1A||L2A)                   L3A=H("LnBranch",L3nonce||L3)
+  
+Assume L1A2A > L3A:
+
+  L1A2A=H("LnBranch",L1A||L2A)          L3A
+                          \            /
+                           v          v
+                Root=H("LnBranch",L3A||L1A2A)
+
+Signature = SIG("lightninginvoicesignature", Root, nodekey)
+```
+
+# Offers
+
+Offers are a precursor to an invoice_request: readers will request an invoice
+(or multiple) based on the offer.  An offer can be much longer-lived than a
+particular invoice, so it has some different characteristics; in particular the amount can be in a non-lightning currency.  It's
+also designed for compactness to fit inside a QR code easily.
+
+Note that the non-signature TLV elements get mirrored into
+invoice_request and invoice messages, so they each have specific and
+distinct TLV ranges.
+
+The human-readable prefix for offers is `lno`.
+
+## TLV Fields for Offers
+
+1. `tlv_stream`: `offer`
+2. types:
+    1. type: 2 (`offer_chains`)
+    2. data:
+        * [`...*chain_hash`:`chains`]
```

Comment:
It seems to me there are many branches created from this value, along with `offer_currency`. The specs could be made easier and cleaner if split into bitcoin-only and alter-chain, maybe a future PR to clean things up.